File: results/2026-04-17_004319/tasks/15-test-results-aggregator/typescript-bun-opus47-1m-medium/generated-code/src/aggregator.ts
// Test result aggregator: parses JUnit XML and JSON test result files,
// aggregates across multiple runs, detects flaky tests, and emits markdown.

export type TestStatus = "passed" | "failed" | "skipped";

export interface TestCase {
  suite: string;
  classname: string;
  name: string;
  status: TestStatus;
  time: number;
  message?: string;
}

export interface TestRun {
  source: string;
  cases: TestCase[];
}

export interface Totals {
  total: number;
  passed: number;
  failed: number;
  skipped: number;
  duration: number;
}

export interface FlakyTest {
  id: string;
  statuses: TestStatus[];
}

export interface Aggregate {
  runs: TestRun[];
  totals: Totals;
  flaky: FlakyTest[];
}

// Parse JUnit XML. Uses a small hand-written parser — keeps zero deps for Bun.
export function parseJUnitXml(xml: string, source = "xml"): TestRun {
  const cases: TestCase[] = [];
  const suiteRegex = /<testsuite\b([^>]*)>([\s\S]*?)<\/testsuite>/g;
  let sMatch: RegExpExecArray | null;
  while ((sMatch = suiteRegex.exec(xml)) !== null) {
    const suiteAttrs = parseAttrs(sMatch[1]);
    const suiteName = suiteAttrs["name"] ?? "unknown";
    const body = sMatch[2];
    // Match <testcase .../> or <testcase ...>...</testcase>
    const caseRegex =
      /<testcase\b([^>]*?)(?:\/>|>([\s\S]*?)<\/testcase>)/g;
    let cMatch: RegExpExecArray | null;
    while ((cMatch = caseRegex.exec(body)) !== null) {
      const attrs = parseAttrs(cMatch[1]);
      const inner = cMatch[2] ?? "";
      let status: TestStatus = "passed";
      let message: string | undefined;
      if (/<failure\b/.test(inner) || /<error\b/.test(inner)) {
        status = "failed";
        const m = /<(?:failure|error)\b([^>]*)/.exec(inner);
        if (m) message = parseAttrs(m[1])["message"];
      } else if (/<skipped\b/.test(inner)) {
        status = "skipped";
      }
      cases.push({
        suite: suiteName,
        classname: attrs["classname"] ?? suiteName,
        name: attrs["name"] ?? "unnamed",
        status,
        time: Number(attrs["time"] ?? "0"),
        message,
      });
    }
  }
  if (cases.length === 0 && !/<testsuite/.test(xml)) {
    throw new Error(`Invalid JUnit XML from '${source}': no <testsuite> found`);
  }
  return { source, cases };
}

function parseAttrs(s: string): Record<string, string> {
  const out: Record<string, string> = {};
  const re = /(\w[\w:-]*)\s*=\s*"([^"]*)"/g;
  let m: RegExpExecArray | null;
  while ((m = re.exec(s)) !== null) out[m[1]] = m[2];
  return out;
}

interface JsonSuite {
  name: string;
  tests: Array<{
    name: string;
    classname?: string;
    status: TestStatus;
    time?: number;
    message?: string;
  }>;
}
interface JsonReport {
  suites: JsonSuite[];
}

export function parseJsonReport(text: string, source = "json"): TestRun {
  let parsed: JsonReport;
  try {
    parsed = JSON.parse(text);
  } catch (e) {
    throw new Error(`Invalid JSON from '${source}': ${(e as Error).message}`);
  }
  if (!parsed || !Array.isArray(parsed.suites)) {
    throw new Error(`Invalid JSON from '${source}': missing 'suites' array`);
  }
  const cases: TestCase[] = [];
  for (const suite of parsed.suites) {
    for (const t of suite.tests ?? []) {
      if (!["passed", "failed", "skipped"].includes(t.status)) {
        throw new Error(
          `Invalid status '${t.status}' for ${suite.name}.${t.name} in '${source}'`,
        );
      }
      cases.push({
        suite: suite.name,
        classname: t.classname ?? suite.name,
        name: t.name,
        status: t.status,
        time: t.time ?? 0,
        message: t.message,
      });
    }
  }
  return { source, cases };
}

// Auto-detect by extension or content.
export async function parseFile(path: string): Promise<TestRun> {
  const file = Bun.file(path);
  if (!(await file.exists())) {
    throw new Error(`Test result file not found: ${path}`);
  }
  const text = await file.text();
  const lower = path.toLowerCase();
  if (lower.endsWith(".xml")) return parseJUnitXml(text, path);
  if (lower.endsWith(".json")) return parseJsonReport(text, path);
  // Fallback: sniff content.
  if (text.trimStart().startsWith("<")) return parseJUnitXml(text, path);
  return parseJsonReport(text, path);
}

export function computeTotals(runs: TestRun[]): Totals {
  const t: Totals = { total: 0, passed: 0, failed: 0, skipped: 0, duration: 0 };
  for (const run of runs) {
    for (const c of run.cases) {
      t.total++;
      t[c.status]++;
      t.duration += c.time;
    }
  }
  // Round duration to 2 decimals to avoid float noise in output.
  t.duration = Math.round(t.duration * 100) / 100;
  return t;
}

// Flaky = a test (classname.name) that both passed and failed across runs.
export function findFlaky(runs: TestRun[]): FlakyTest[] {
  const map = new Map<string, TestStatus[]>();
  for (const run of runs) {
    for (const c of run.cases) {
      const id = `${c.classname}.${c.name}`;
      const arr = map.get(id) ?? [];
      arr.push(c.status);
      map.set(id, arr);
    }
  }
  const flaky: FlakyTest[] = [];
  for (const [id, statuses] of map) {
    if (statuses.includes("passed") && statuses.includes("failed")) {
      flaky.push({ id, statuses });
    }
  }
  flaky.sort((a, b) => a.id.localeCompare(b.id));
  return flaky;
}

export function aggregate(runs: TestRun[]): Aggregate {
  return { runs, totals: computeTotals(runs), flaky: findFlaky(runs) };
}

export function renderMarkdown(agg: Aggregate): string {
  const { totals, flaky, runs } = agg;
  const passRate =
    totals.total === 0
      ? "0.0"
      : ((totals.passed / totals.total) * 100).toFixed(1);
  const status = totals.failed === 0 ? "✅ PASS" : "❌ FAIL";
  const lines: string[] = [];
  lines.push(`# Test Results ${status}`);
  lines.push("");
  lines.push(`**Runs aggregated:** ${runs.length}`);
  lines.push("");
  lines.push("## Totals");
  lines.push("");
  lines.push("| Metric | Value |");
  lines.push("| --- | --- |");
  lines.push(`| Total | ${totals.total} |`);
  lines.push(`| Passed | ${totals.passed} |`);
  lines.push(`| Failed | ${totals.failed} |`);
  lines.push(`| Skipped | ${totals.skipped} |`);
  lines.push(`| Duration | ${totals.duration.toFixed(2)}s |`);
  lines.push(`| Pass rate | ${passRate}% |`);
  lines.push("");
  lines.push("## Flaky Tests");
  lines.push("");
  if (flaky.length === 0) {
    lines.push("_No flaky tests detected._");
  } else {
    lines.push("| Test | Results |");
    lines.push("| --- | --- |");
    for (const f of flaky) {
      lines.push(`| ${f.id} | ${f.statuses.join(", ")} |`);
    }
  }
  lines.push("");
  return lines.join("\n");
}

export async function main(argv: string[]): Promise<number> {
  const files = argv.filter((a) => !a.startsWith("--"));
  if (files.length === 0) {
    console.error("Usage: bun run src/aggregator.ts <file1> [file2...]");
    return 2;
  }
  const runs: TestRun[] = [];
  for (const f of files) runs.push(await parseFile(f));
  const agg = aggregate(runs);
  const md = renderMarkdown(agg);
  const outPath = process.env.GITHUB_STEP_SUMMARY;
  if (outPath) await Bun.write(outPath, md);
  console.log(md);
  return agg.totals.failed > 0 ? 1 : 0;
}

if (import.meta.main) {
  const code = await main(Bun.argv.slice(2));
  process.exit(code);
}

File: results/2026-04-17_004319/tasks/15-test-results-aggregator/default-opus47-1m-high/generated-code/aggregator.py
"""
Test-results aggregator.

Parses JUnit XML and JSON test result files produced by a matrix build,
combines them into a single summary, identifies flaky tests (any test
that both passed and failed across the matrix), and renders a Markdown
report suitable for a GitHub Actions job summary.

Usage:
    python aggregator.py <dir> [--output SUMMARY.md] [--fail-on-failures]

If --output is omitted and $GITHUB_STEP_SUMMARY is set (i.e. inside a
GitHub Actions job) the report is appended there as well.
"""
from __future__ import annotations

import argparse
import json
import os
import sys
import xml.etree.ElementTree as ET
from dataclasses import dataclass, field
from pathlib import Path
from typing import Iterable, Sequence


# ---------------------------------------------------------------------------
# Data types
# ---------------------------------------------------------------------------


VALID_STATUSES = {"passed", "failed", "skipped"}


@dataclass(frozen=True)
class TestCase:
    """A single test case result."""

    name: str
    classname: str
    status: str  # "passed" | "failed" | "skipped"
    duration: float
    message: str | None = None

    # Tell pytest not to collect this dataclass as a test class.
    __test__ = False

    def __post_init__(self) -> None:
        if self.status not in VALID_STATUSES:
            raise ValueError(f"Invalid status {self.status!r}")


@dataclass
class TestRun:
    """The parsed contents of one result file (one matrix shard)."""

    suite_name: str
    cases: list[TestCase]
    source: str = ""

    __test__ = False


@dataclass
class Aggregate:
    """Summed totals over all runs."""

    runs: list[TestRun]
    total: int
    passed: int
    failed: int
    skipped: int
    duration: float


@dataclass
class FlakyEntry:
    """A test that flipped between passed and failed across runs."""

    name: str
    classname: str
    passed_count: int
    failed_count: int


# ---------------------------------------------------------------------------
# JUnit XML parsing
# ---------------------------------------------------------------------------


def _status_from_junit(testcase: ET.Element) -> tuple[str, str | None]:
    """Return (status, message) for a JUnit <testcase> element."""
    if testcase.find("failure") is not None:
        el = testcase.find("failure")
        return "failed", (el.get("message") if el is not None else None)
    if testcase.find("error") is not None:
        el = testcase.find("error")
        return "failed", (el.get("message") if el is not None else None)
    if testcase.find("skipped") is not None:
        return "skipped", None
    return "passed", None


def parse_junit_xml(path: Path) -> TestRun:
    """Parse a JUnit XML file into a TestRun.

    Handles both a single <testsuite> root and a <testsuites> wrapper.
    """
    path = Path(path)
    if not path.exists():
        raise FileNotFoundError(f"JUnit file not found: {path}")
    try:
        tree = ET.parse(path)
    except ET.ParseError as exc:
        raise ValueError(f"Failed to parse JUnit XML {path}: {exc}") from exc

    root = tree.getroot()
    suites: list[ET.Element]
    if root.tag == "testsuites":
        suites = list(root.findall("testsuite"))
    elif root.tag == "testsuite":
        suites = [root]
    else:
        raise ValueError(
            f"Unexpected root element <{root.tag}> in {path}; "
            "expected <testsuite> or <testsuites>"
        )

    cases: list[TestCase] = []
    # Name the run after the first <testsuite>; good enough for grouping.
    suite_name = suites[0].get("name", path.stem) if suites else path.stem
    for suite in suites:
        for tc in suite.findall("testcase"):
            status, message = _status_from_junit(tc)
            try:
                duration = float(tc.get("time", "0") or 0)
            except ValueError:
                duration = 0.0
            cases.append(
                TestCase(
                    name=tc.get("name", "<unnamed>"),
                    classname=tc.get("classname", suite.get("name", "")),
                    status=status,
                    duration=duration,
                    message=message,
                )
            )

    return TestRun(suite_name=suite_name, cases=cases, source=str(path))


# ---------------------------------------------------------------------------
# JSON parsing
# ---------------------------------------------------------------------------


def parse_json(path: Path) -> TestRun:
    """Parse a JSON test-results file.

    Expected shape:
      { "suite": "name", "tests": [
          {"name": str, "classname": str, "status": str,
           "duration": number, "message": str?}
      ] }
    """
    path = Path(path)
    if not path.exists():
        raise FileNotFoundError(f"JSON file not found: {path}")
    try:
        data = json.loads(path.read_text())
    except json.JSONDecodeError as exc:
        raise ValueError(f"Failed to parse JSON {path}: {exc}") from exc

    if "tests" not in data:
        raise ValueError(f"JSON file {path} missing 'tests' list")

    cases: list[TestCase] = []
    for entry in data["tests"]:
        status = entry.get("status", "passed")
        cases.append(
            TestCase(
                name=entry.get("name", "<unnamed>"),
                classname=entry.get("classname", ""),
                status=status,
                duration=float(entry.get("duration", 0) or 0),
                message=entry.get("message"),
            )
        )
    return TestRun(
        suite_name=data.get("suite", path.stem),
        cases=cases,
        source=str(path),
    )


# ---------------------------------------------------------------------------
# Dispatcher
# ---------------------------------------------------------------------------


def parse_file(path: Path) -> TestRun:
    """Dispatch to the right parser based on extension."""
    path = Path(path)
    ext = path.suffix.lower()
    if ext == ".xml":
        return parse_junit_xml(path)
    if ext == ".json":
        return parse_json(path)
    raise ValueError(f"Unsupported file extension {ext!r} for {path}")


# ---------------------------------------------------------------------------
# Aggregation
# ---------------------------------------------------------------------------


def aggregate(runs: Sequence[TestRun]) -> Aggregate:
    total = passed = failed = skipped = 0
    duration = 0.0
    for r in runs:
        for c in r.cases:
            total += 1
            duration += c.duration
            if c.status == "passed":
                passed += 1
            elif c.status == "failed":
                failed += 1
            elif c.status == "skipped":
                skipped += 1
    return Aggregate(
        runs=list(runs),
        total=total,
        passed=passed,
        failed=failed,
        skipped=skipped,
        duration=duration,
    )


def find_flaky(runs: Sequence[TestRun]) -> list[FlakyEntry]:
    """Return tests that both passed and failed across the given runs."""
    counts: dict[tuple[str, str], dict[str, int]] = {}
    for r in runs:
        for c in r.cases:
            key = (c.classname, c.name)
            counts.setdefault(key, {"passed": 0, "failed": 0, "skipped": 0})
            counts[key][c.status] += 1

    flaky: list[FlakyEntry] = []
    for (classname, name), tally in counts.items():
        if tally["passed"] > 0 and tally["failed"] > 0:
            flaky.append(
                FlakyEntry(
                    name=name,
                    classname=classname,
                    passed_count=tally["passed"],
                    failed_count=tally["failed"],
                )
            )
    flaky.sort(key=lambda e: (e.classname, e.name))
    return flaky


# ---------------------------------------------------------------------------
# Markdown rendering
# ---------------------------------------------------------------------------


def _status_icon(status: str) -> str:
    return {"passed": "PASS", "failed": "FAIL", "skipped": "SKIP"}.get(status, status)


def generate_markdown(agg: Aggregate, flaky: list[FlakyEntry]) -> str:
    lines: list[str] = []
    lines.append("# Test Results Summary")
    lines.append("")
    lines.append("## Totals")
    lines.append("")
    lines.append("| Metric | Count |")
    lines.append("| --- | --- |")
    lines.append(f"| Total | {agg.total} |")
    lines.append(f"| Passed | {agg.passed} |")
    lines.append(f"| Failed | {agg.failed} |")
    lines.append(f"| Skipped | {agg.skipped} |")
    lines.append(f"| Duration (s) | {agg.duration:.2f} |")
    lines.append("")

    lines.append("## Runs")
    lines.append("")
    lines.append("| Suite | Source | Passed | Failed | Skipped | Duration (s) |")
    lines.append("| --- | --- | --- | --- | --- | --- |")
    for r in agg.runs:
        p = sum(1 for c in r.cases if c.status == "passed")
        f = sum(1 for c in r.cases if c.status == "failed")
        s = sum(1 for c in r.cases if c.status == "skipped")
        d = sum(c.duration for c in r.cases)
        source = Path(r.source).name if r.source else ""
        lines.append(f"| {r.suite_name} | {source} | {p} | {f} | {s} | {d:.2f} |")
    lines.append("")

    # Failures detail
    failures = [
        (r, c) for r in agg.runs for c in r.cases if c.status == "failed"
    ]
    if failures:
        lines.append("## Failures")
        lines.append("")
        for r, c in failures:
            msg = c.message or "(no message)"
            lines.append(f"- **{c.classname}::{c.name}** ({r.suite_name}) — {msg}")
        lines.append("")

    # Flaky
    lines.append("## Flaky Tests")
    lines.append("")
    if flaky:
        lines.append("| Test | Passed | Failed |")
        lines.append("| --- | --- | --- |")
        for entry in flaky:
            lines.append(
                f"| {entry.classname}::{entry.name} | {entry.passed_count} | {entry.failed_count} |"
            )
    else:
        lines.append("_No flaky tests detected._")
    lines.append("")

    return "\n".join(lines)


# ---------------------------------------------------------------------------
# CLI
# ---------------------------------------------------------------------------


def _discover(directory: Path) -> list[Path]:
    """Return sorted list of .xml and .json files directly in `directory`
    and its subdirectories."""
    patterns = ("*.xml", "*.json")
    found: list[Path] = []
    for pat in patterns:
        found.extend(sorted(directory.rglob(pat)))
    return sorted(found)


def run_cli(argv: list[str]) -> int:
    parser = argparse.ArgumentParser(
        description="Aggregate JUnit XML / JSON test results into a Markdown summary.",
    )
    parser.add_argument("input_dir", help="Directory containing result files (.xml/.json)")
    parser.add_argument("--output", "-o", help="Write the Markdown report to this file")
    parser.add_argument(
        "--fail-on-failures",
        action="store_true",
        help="Exit with code 1 if any test failed",
    )
    args = parser.parse_args(argv)

    input_dir = Path(args.input_dir)
    if not input_dir.exists() or not input_dir.is_dir():
        print(f"error: input directory does not exist: {input_dir}", file=sys.stderr)
        return 2

    files = _discover(input_dir)
    if not files:
        print(f"warning: no result files found in {input_dir}", file=sys.stderr)

    runs: list[TestRun] = []
    for f in files:
        try:
            runs.append(parse_file(f))
        except (ValueError, FileNotFoundError) as exc:
            print(f"error parsing {f}: {exc}", file=sys.stderr)
            return 3

    agg = aggregate(runs)
    flaky = find_flaky(runs)
    md = generate_markdown(agg, flaky)

    # Print to stdout (visible in act logs).
    print(md)

    # Optional explicit output file
    if args.output:
        Path(args.output).write_text(md)

    # Also write to $GITHUB_STEP_SUMMARY when present.
    step_summary = os.environ.get("GITHUB_STEP_SUMMARY")
    if step_summary:
        with open(step_summary, "a") as fh:
            fh.write(md)
            fh.write("\n")

    if args.fail_on_failures and agg.failed > 0:
        return 1
    return 0


def main() -> int:  # pragma: no cover - thin wrapper
    return run_cli(sys.argv[1:])


if __name__ == "__main__":
    sys.exit(main())
